pingcap/tidb
# ddl: support `ALTER PLACEMENT` clause

### What problem does this PR solve?

It is a part of the project `Define Placement Rules in SQL`. This PR adds the basic support for `ALTER PLACEMENT`. It is split into several commits to clarify what it has done exactly.

1. check the count on TiDB now(https://github.com/pingcap/parser/pull/965).
2. alter should fail all previous related statements
3. id format is needed to change to support `ALTER`
4. local placement types definition(https://github.com/pingcap/pd/issues/2701), which eliminates one leak test ignore option
5. after 4 commits, it is super easy to implement `ALTER`
6. some tests and fixes

### Check List

Tests

- Unit test
- Integration test
- Manual test

Start `pd-server`, `tikv-server`, `tidb-server`(PD is excepted to have newer version than a26d9cf), enable placement rules using `pd-ctl`.

And the default rule from PD is not considered in the tests, verify the following cases using `pd-ctl`:

1. `ALTER TABLE t ALTER PARTITION p0 ADD PLACEMENT POLICY
CONSTRAINTS='["+zone=sh"]' ROLE=voter REPLICAS=3`. It is excepted to see one more rule using `pd-ctl config placement-rules show`.
2. `ALTER TABLE t ALTER PARTITION p0 ADD PLACEMENT POLICY
CONSTRAINTS='{"+zone=sh":1, "-zone=bj":2}' ROLE=voter REPLICAS=6`. It is excepted to see three more rules with 1,2,3 replicas.
3. `ALTER TABLE t ALTER PARTITION p0 ALTER PLACEMENT POLICY
CONSTRAINTS='{"+zone=sh":4, "-zone=bj":5}' ROLE=voter REPLICAS=10`. It is excepted to see three rules(only three rules) with 4,5,1 replicas.

### Release note

- support `ALTER PLACEMENT` clause
